Lightweight task scheduler

gavinandresen opened the pull request:
Threads have noticeable overhead on 32-bit systems, so using fewer threads for simple tasks is better than starting up threads willy-nilly.

The first commit implements a simple task queue class (CScheduler) which can be serviced by one or more threads.

The second and third port the networking code's DumpAddresses thread to be a task instead.

Porting other threads (ThreadFlushWalletDB is a good candidate) is left as an exercise for future pull requests.

The CScheduler class was developed and tested with code here: https://github.com/gavinandresen/scheduler  (see main.cpp)

I tested the DumpAddresses change running with -debug=net and a hacked DEBUG_ADDRESS_INTERVAL, and made sure  I got more than one "Flushed %d addresses to peers.dat" in my debug.log.

Commit:
[Qt] add defaultConfirmTarget constant to sendcoinsdialog

- replaces some hard-coded values for the default confirmation target
- also simplify code that is using the new constant

gavinandresen committed:
CScheduler class for lightweight task scheduling

Simple class to manage a task queue that is serviced by one or
more threads.

gavinandresen committed:
Use CScheduler for net's DumpAddresses

Instead of starting Yet Another Thread to dump addresses,
use CScheduler to do it.

sipa commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,87 @@
+            // Wait until either there is a new task, or until
+            // the time of the first item on the queue:
+            while (!taskQueue.empty() &&
+                   newTaskScheduled.wait_until(lock, taskQueue.begin()->first) != boost::cv_status::timeout)
```
Does this work when the timestamp of the first item in the queue is in the past?

sipa commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,87 @@
+    while (1)
+    {
+        if (taskQueue.empty())
+        {
```
Nit: coding style (braces on the same line).

sipa commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,87 @@
+        boost::unique_lock<boost::mutex> lock(newTaskMutex);
+        taskQueue.insert(std::make_pair(t, f));
+    }
+    newTaskScheduled.notify_all();
```
Nit: notify_one should be enough (and slightly more efficient, if we'd ever have multiple runner threads).

jgarzik commented:
concept ACK
CScheduler ACK

* Yes, DumpAddresses thread is a prime candidate for this
* Ditto ThreadFlushWalletDB
* Disagree on ThreadImport -- ThreadImport is a prime candidate for something to be executed in a long running thread of its own.

sipa commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,87 @@
+        if (taskQueue.empty())
+        {
+            // Wait until there is something to do.
+            newTaskScheduled.wait(lock);
```
.wait() may wake spuriously, so it's better to turn the surrounding if into a while loop. I believe that means you also won't have to deal with the queue-empty case in the loop below.

sipa commented on src/scheduler.h:
```diff
@@ -0,0 +1,66 @@
+{
+public:
+    CScheduler();
+
```
Unsure about this, but leaving the thread running forever may cause destruction order issues if a task is running exactly at shutdown. A destructor that sets an 'exit' bool flag (which the scheduler thread checks), and waits for all scheduler threads to exit may be needed.

gavinandresen replied:
Thread lifetime managment is up to the caller. For Bitcoin, the thread is added to the main thread group, and all the threads are interrupted and joined before shutdown (see WaitForShutdown() in bitcoind.cpp or BitcoinCore::shutdown() in qt/bitcoin.cpp).

sipa commented:
Agree with @jgarzik; threadimport should not run in the scheduler, as it is expected to take up to hours - blocking several other operations. Wallet flushing is a good candidate.

gavinandresen committed:
Create a scheduler thread for lightweight tasks

gavinandresen commented:
Fixed @sipa's nits, including running scheduler.cpp/.h through clang-format.

And reverted the ThreadImport change, so it is it's own thread again.

theuni commented:
```boost::condition_variable::wait_until``` and ```boost::cv_status``` didn't show up until 1.50. Can they be worked around?

Otherwise, we'll need to enforce >= 1.50 as a build requirement.

sipa commented:
They're easy to work with using timed_wait, I think.

jgarzik commented:
ut ACK latest rev

gavinandresen commented:
Three changes:

1. I decided I didn't like the the asymmetry of CScheduler creating the thread, but the caller managing it. So I replaced CScheduler::run() with CScheduler::ServiceQueue, which the caller thread-ifys. And added code and a destructor that asserts if thread management isn't done correctly. That made it easy to wrap the scheduler trhead in util.h's TraceThread, to get debug.log tracing and handling of exceptions that might be thrown from the scheduled tasks.

2. Added compatibility code for boost version 1.49 and below.

3. I fixed a race-condition bug when there are multiple threads servicing the queue.

I think I'm all done tweaking this now.

sipa commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,104 @@
+#endif
+            // If there are multiple threads, the queue can empty while we're waiting (another
+            // thread may service the task we were waiting on).
+            if (taskQueue.empty()) continue;
```
I was wrong earlier. If you want to support multiple runner threads, you need to do this check inside the while loops above, as the item can disappear while they wait.

gavinandresen replied:
Mmm. I convinced myself an empty check isn't necessary there, but belt&suspenders says I should just put it back.

(not needed because: if the queue is emptied during timeout, then the wait loop is timed out and exited and the empty queue is caught here. And since there is no unschedule-task, the only other way the wait can exit is if a new task is added to the queue, which wakes up only one thread and so the queue can't be empty).

I'll add the empty condition to the wait loops.

sipa replied:
Ok, you convinced me... but indeed, belt & suspenders.

You could even combine the two loops: while (taskQueue.empty() || wait_until(taskQueue.begin()->first)) ?

sipa commented on src/scheduler.h:
```diff
@@ -0,0 +1,74 @@
+    // Start a thread to service the task queue.
+    // Run may be called multiple times to have multiple
+    // threads servicing the queue.
+    boost::thread* run();
```
This doesn't seem to exist anymore.

sipa commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,104 @@
+            // Unlock before calling f, so it can reschedule itself or another task
+            // without deadlocking:
+            lock.unlock();
+            f();
```
Do you want the loop to exit if the user function throws an exception?

gavinandresen replied:
If it is a thread_interrupted exception, definitely want to exit.

Any other exception, it seems best to leave it up to the calling code what to do.

For Bitcoin, the TraceThread will print any exceptions (other than thread_interrupted) to debug.log and exit the scheduler thread. I think that is the behavior we want -- tasks should catch their own exceptions, if they don't that is a bug.

I don't feel strongly, though-- instead of wrapping with TraceThread we could bring back LoopForever so uncaught exceptions from tasks are logged to debug.log but the scheduler loop keeps running.

sipa replied:
Fair enough. I mostly commented because it may make debugging a bit harder, but I agree: tasks should not throw exceptions (except thread_interrupted, perhaps).

laanwj replied:
As long as run-away exceptions are logged, any solution here is fine with me. It is indeed a bug if it happens.

sipa commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,104 @@
+
+void CScheduler::Repeat(CScheduler::Function f, int64_t deltaSeconds)
+{
+    f();
```
(nit) You may not want to run f inside the direct call path, as that means it will run while holding the locks the caller has (typically you expect scheduled tasks to run free from any locks).

sipa replied:
Nevermind, this isn't the public method.

sipa commented:
extremely-lightly-tested ACK (as in: does not crash).

gavinandresen commented:
Final (?) nits picked (got rid of the obsolete run from the .h, added belt&suspenders checks for empty queue, ran through clang-format again), this should be ready for merge.

I also did some more testing, writing a stress-test that created 4,000 tasks serviced by 100 threads over a couple of milliseconds ( https://github.com/gavinandresen/scheduler/blob/master/main.cpp ), and all is well, including running under valgrind.

jgarzik commented:
lightly tested ACK

theuni committed:
build: make libboost_chrono mandatory

previously it was only used with certain boost versions. Now all versions
require it.

theuni committed:
scheduler: fix with boost <= 1.50

theuni commented:
@gavinandresen See the top 2 commits here: https://github.com/theuni/bitcoin/tree/5964

Without the timer fix, using your test program I was flooded with "Repeat!" and everything else returned immediately. With the change from https://github.com/theuni/bitcoin/commit/4d0819394249be25c34db28e4c3648a71ab238fa I get:
```
Microtask counts: 190 198 199 197 196 197 205 201 217 200 
Sum: 2000
Wait... negative?
One
Two
AlsoTwo
Three
Gonna start repeating every 2 secs
Five
Repeat!
Repeat!
Repeat!
Eleven!
Repeat!
Long-running task, gonna sleep for 11seconds
One
Repeat!
Two
Repeat!
Repeat!
Repeat!
Repeat!
Done sleeping
Repeat!
```
Each looks to be taking as long as they should.

It'd be great to see that moved into a test. Given boost's wait() problems with 1.50-1.52 iirc, I think there's reason to be cautious.

gavinandresen commented:
@theuni : thanks for testing, I'll pull in those commits and I'll create a unit test with the microtask test -- does it fail without your fixes?

I don't want to create a unit test that takes 20 seconds to run.

gavinandresen committed:
CScheduler unit test

gavinandresen commented:
Pulled in fixes from @theuni, and added a new unit test. I measured run-time and memory usage before and after the new unit test, no significant difference on my 64-bit OSX machine:

before:
```
Running 148 test cases...

*** No errors detected
        6.57 real         4.29 user         0.47 sys
 197120000  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     49901  page reclaims
         0  page faults
         0  swaps
         0  block input operations
       486  block output operations
         0  messages sent
         0  messages received
         0  signals received
       257  voluntary context switches
      4137  involuntary context switches
```
after:
```
Running 148 test cases...

*** No errors detected
        6.59 real         4.39 user         0.53 sys
 197238784  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     50350  page reclaims
         0  page faults
         0  swaps
         0  block input operations
       434  block output operations
         0  messages sent
         0  messages received
         0  signals received
       248  voluntary context switches
     28623  involuntary context switches
```

sipa commented:
utACK

pstratem commented:
There should be a mechanism for the scheduled Function to specify the next time it's called.

This is necessary for high precision random intervals.

gavinandresen commented:
Functions can reschedule themselves-- pass a reference to the CScheduler object to the function (via boost::bind and boost::ref). See the CScheduler unit tests for examples.

sipa commented on src/test/scheduler_tests.cpp:
```diff
@@ -0,0 +1,97 @@
+                                             randomDelta(rng), tReschedule);
+        microTasks.schedule(f, t);
+    }
+    boost::this_thread::sleep_for(boost::chrono::microseconds(600));
```
This line doesn't compile for me.

gavinandresen replied:
I'll implement a similar fix to utiltime.cpp's MilliSleep to workaround broken versions of boost.

sipa commented:
Tested ACK.

gavinandresen commented:
Rebased to fix a trivial merge conflict (#include removed from src/bitcoind.cpp). I think this should be merged for 0.11.

luke-jr commented:
```
test/scheduler_tests.cpp: In member function ‘void scheduler_tests::manythreads::test_method()’:
test/scheduler_tests.cpp:68:5: error: ‘sleep_for’ is not a member of ‘boost::this_thread’
test/scheduler_tests.cpp:84:5: error: ‘sleep_for’ is not a member of ‘boost::this_thread’
```

gavinandresen commented:
@luke-jr : fixed. I'd fixed that before, I think it got lost in a rebase / shuffle between working on home/work trees.

The fix is in the unit test commit, and is:
```
static void MicroSleep(uint64_t n)
{
#if defined(HAVE_WORKING_BOOST_SLEEP_FOR)
    boost::this_thread::sleep_for(boost::chrono::microseconds(n));
#elif defined(HAVE_WORKING_BOOST_SLEEP)
    boost::this_thread::sleep(boost::posix_time::microseconds(n));
#else
    //should never get here
    #error missing boost sleep implementation
#endif
}
```

laanwj commented on src/util.h:
```diff
@@ -203,43 +203,6 @@ void SetThreadPriority(int nPriority);
- *    boost::function<void()> f = boost::bind(&FunctionWithArg, argument);
- *    threadGroup.create_thread(boost::bind(&LoopForever<boost::function<void()> >, "nothing", f, milliseconds));
- */
-template <typename Callable> void LoopForever(const char* name,  Callable func, int64_t msecs)
```
Happy to see that this gets rid of the awful `LoopForever()` in util.h.

laanwj commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,98 @@
+static void Repeat(CScheduler* s, CScheduler::Function f, int64_t deltaSeconds)
+{
+    f();
+    s->scheduleFromNow(boost::bind(&Repeat, s, f, deltaSeconds), deltaSeconds);
```
For repeating tasks, there are two options:

1. Schedule the new run deltaSeconds after completion of the task, every time
2. Schedule every deltaSeconds, no matter how long the task takes (unless the task takes longer than the interval, then something special has to be done)

So 1 schedules from the end time of the task, 2 schedules from the beginning time of the task,.

This appears to implement 1. Good enough in most cases (we don't need clockwork precision) but this may be worth documenting.

laanwj commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,98 @@
+
+// wait_until needs boost 1.50 or later; older versions have timed_wait:
+#if BOOST_VERSION < 105000
+            while (!taskQueue.empty() && newTaskScheduled.timed_wait(lock, toPosixTime(taskQueue.begin()->first)) == true) {
```
nit: please leave out the `== true`, booleans already make good truth values

laanwj commented:
ACK apart from the nit

gavinandresen commented:
Nits picked, will merge as soon as Travis gives the green light.

gavinandresen commented:
Travis was unhappy-- one of the builds was taking more than 50 minutes to complete.

I toned down the unit test by a factor of ten, in case that is what was causing the problem (it now services hundreds of tasks using 10 threads, instead of thousands using 100 threads).... if the builds STILL take longer than normal I'll have to recruit @theuni to help debug.

laanwj commented:
@gavinandresen It initially failed; so I merged #6136 and tried again. It's expected to take somewhat longer but should at least succeed.

theuni commented:
@gavinandresen I think it's partially bad timing. https://github.com/bitcoin/bitcoin/pull/6136 forced travis to rebuild all dependencies and build up a new cache. Should be back to normal in an hour or so.

btcdrak commented on src/scheduler.cpp:
```diff
@@ -0,0 +1,98 @@
+                newTaskScheduled.wait(lock);
+            }
+// Wait until either there is a new task, or until
+// the time of the first item on the queue:
```
tiny nit, is the indenting here off?

gavincoin replied:
the comment is for the #if so the indentation is correct

luke-jr commented:
Build and unit tests pass on Debian 7: ACK

theuni commented:
@gavinandresen I got this earlier on an unrelated PR, looks very related/suspicious: https://travis-ci.org/bitcoin/bitcoin/jobs/62615004#L1344 .

gavinandresen commented:
Hmmm... Looks like the threads were interrupted before finishing (slow VM maybe). I'll try to rework the test tomorrow so it doesn't interrupt the threads.
